apple/testing/default_runner/macos_test_runner.bzl

```diff
@@ -14,38 +14,40 @@
 
 """Sample Apple test runner rule."""
 
-load(
-    "@build_bazel_apple_support//lib:xcode_support.bzl",
-    "xcode_support",
-)
 load(
     "@build_bazel_rules_apple//apple/testing:apple_test_rules.bzl",
     "AppleTestRunnerInfo",
 )
 
 def _get_xctestrun_template_substitutions(ctx):
     """Returns the template substitutions for the xctestrun template."""
-    xcode_version = ctx.attr._xcode_config[apple_common.XcodeVersionConfig]
+    XCTESTRUN_INSERT_LIBRARIES = "%(xctestrun_insert_libraries)s"
+    substitutions = {
+        XCTESTRUN_INSERT_LIBRARIES: "",
+    }
+
+    xcode_version = ctx.attr._xcode_config[apple_common.XcodeVersionConfig].xcode_version()
+
+    # XCode version is not present on non-macOS platforms
+    if not xcode_version:
+        return substitutions
```

Comment:
This would hide the actual failure if you build on macOS _and_ Xcode isn't present.

Also, `xcode_version()` can be present on non-macOS platforms with a cross-compiling toolchain.

I'm not sure how to deal with this 😢

Reply:
I think, it shouldn't hide the failure when XCode isn't present on macOS. I would expect that this change just moves the failure to the Execution Phase. I assume the build rule will look for an clang/llvm binary and won't find it. I could try to test this by uninstalling XCode but it would probably useful to capture this in a CI environment too.

Good point on `xcode_version()` being present during cross-platform compilation. Is any of these rules used for cross-compiling to other platfroms? If not, I think, these rules will be just skipped and it should be safe.

Reply:
https://github.com/apple-cross-toolchain/rules_applecross

Reply:
This looks useful. Thanks for sharing, @brentleyjones ! I'll give it a try.

Out of curiosity: what are your thoughts around using `target_compatible_with`? Some people might opt for not setting up a toolchain on Linux.

Reply:
It would be useful to be able to use `target_compatible_with`, but we need to be able to support cross-compiling and when Xcode isn't available on macOS.

Reply:
It's probably my lack of experience with rules_apple, so I'll try to summaries the cases to consider:

- macOS + XCode
- macOS and no XCode
- non-macOS platform + XCode
- non-macOS platform and no XCode

`if not xcode_version:` should not change the current logic as the `is_xcode_at_least_version()` contains the following:

```starlark
    current_version = xcode_config.xcode_version()
    if not current_version:
        fail("Could not determine Xcode version at all. This likely means Xcode isn't available; " +
             "if you think this is a mistake, please file an issue.")
```

This change should just inline the [is_xcode_at_least_version()](https://github.com/bazelbuild/apple_support/blob/master/lib/xcode_support.bzl#L17) function and make it graceful by returning an 'empty' dictionary.

Only difference I can see that if the target is actually executed, it fails at execution phase and not analysis phase. I'm not sure what you mean by hiding the error, because I would still expect the build to fail if XCode is not present, just at a different time and perhaps with a different reason. I think, there a potential to improve developer experience and giving clear feedback on missing XCode by doing a check in `macos_test_runner.template.sh`. What do you think about that approach? Would that work?

Reply:
Yes, I'm aware that this runner can be overridden. I experimented with adding a `select()` and using the `//test/testdata/rules:dummy_runner` on Linux. However, this file still gets loaded and analysed by Bazel and that makes the build fail. My understanding is because this particular code executes during the creation of the `test_runner_rule()`, so it happens before it's even used but please correct me if I misunderstood this.

I've reverted the refactoring. 👍

Reply:
What I mean was:

1. Copy this test runner rule into your repo, for example into `rules/macos_test_runner.bzl`.
2. Make changes to it as you like.
3. Create a new runner target using your new rule, for example `//:custom_macos_runner`.
4. Pass the new runner to your test target's `runner` attribute.

```starlark
# BUILD
load("//rules:macos_test_runner.bzl", "macos_test_runner")

macos_test_runner(
    name = "custom_macos_runner",
)

# ...

macos_unit_test(
    name = "ButtonsMacLogicTests",
    bundle_id = "com.google.logic",
    minimum_os_version = "10.10",
    deps = [":ButtonsMacTestsLib"],
    runner = ":custom_macos_runner",
    target_compatible_with = ["@platforms//os:macos"],
)
```

This way, the default runner won't be evaluated in the analysis phase anymore.